src/sagemaker_xgboost_container/algorithm_mode/train.py

```diff
@@ -203,14 +201,100 @@ def train_job(train_cfg, train_dmatrix, val_dmatrix, train_val_dmatrix, model_di
     else:
         train_cfg.pop('eval_metric', None)
 
-    # Set callback evals
-    watchlist = [(train_dmatrix, 'train')]
-    if val_dmatrix is not None:
-        watchlist.append((val_dmatrix, 'validation'))
+    logging.info("Train matrix has {} rows and {} columns".format(train_dmatrix.num_row(), train_dmatrix.num_col()))
+    if val_dmatrix:
+        logging.info("Validation matrix has {} rows".format(val_dmatrix.num_row()))
+
+    try:
+        kfold = train_cfg.pop("_kfold", None)
+
+        if kfold is None:
+            xgb_model, iteration, callbacks, watchlist = get_callbacks_watchlist(
+                train_cfg, train_dmatrix, val_dmatrix, model_dir, checkpoint_dir)
+            add_debugging(callbacks=callbacks, hyperparameters=train_cfg, train_dmatrix=train_dmatrix,
+                          val_dmatrix=val_dmatrix)
+
+            bst = xgb.train(train_cfg, train_dmatrix, num_boost_round=num_round-iteration, evals=watchlist,
+                            feval=configured_feval, early_stopping_rounds=early_stopping_rounds,
+                            callbacks=callbacks, xgb_model=xgb_model, verbose_eval=False)
+
+        else:
+            logging.info("Run {}-fold cross validation with {} rows".format(kfold, train_val_dmatrix.num_row()))
+
+            bst = []
+            evals_results = []
+
+            num_cv_round = train_cfg.pop("_num_cv_round", 1)
+            for cv_round in range(num_cv_round):
```

Comment:
this is difficult to follow - we need an explanation on what's a 'round' vs 'fold' and why we need both.

Reply:
The 'round' is about how many times we run cross validation. Therefore, it is known as repeated cross validation (we reshuffle data before each round). The purpose is to generate sufficient number of models for ensemble. The benchmarks show that ensemble of 3 rounds of cross validation models increase the accuracy significantly.

The fold refers the each training inside of one cross validation.

Reply:
My concern is not the repeated cross-validation but rather using the correlated models to build a stack of XGBoost models for inference. I'm not convinced that is a good approach and am looking for scientific papers that provide the theory and results. I strongly believe that the accuracy you will achieve will be same if you increase complexity of base XGBoost model to match the model complexity introduced by combining multiple models.

Reply:
[AutoGluon](https://arxiv.org/abs/2003.06505) does ensemble with repeated cross validation with XGBoost, lightGBM, RandomForest and a few other algorithms. Their paper shows significant accuracy improvement with this strategy even though XGBoost, lightGBM and RandomForest are all ensemble algorithm already.